src/stacktrace.coffee

```diff
@@ -0,0 +1,78 @@
+{ getSourceMap } = require "./sourcemap"
+
+attach = (stackRoot) ->
+  # Based on http://v8.googlecode.com/svn/branches/bleeding_edge/src/messages.js
+  # Modified to handle sourceMap
+  formatSourcePosition = (frame, getSourceMapping) ->
+    filename = undefined
+    fileLocation = ''
+
+    if frame.isNative()
+      fileLocation = "native"
+    else
+      if frame.isEval()
+        filename = frame.getScriptNameOrSourceURL()
+        fileLocation = "#{frame.getEvalOrigin()}, " unless filename
+      else
+        filename = frame.getFileName()
+
+      filename or= "<anonymous>"
+
+      line = frame.getLineNumber()
+      column = frame.getColumnNumber()
+
+      # Check for a sourceMap position
+      source = getSourceMapping filename, line, column
+      fileLocation =
+        if source
+          "#{filename}:#{source[0]}:#{source[1]}"
+        else
+          "#{filename}:#{line}:#{column}"
+
+    functionName = frame.getFunctionName()
+    isConstructor = frame.isConstructor()
+    isMethodCall = not (frame.isToplevel() or isConstructor)
+
+    if isMethodCall
+      methodName = frame.getMethodName()
+      typeName = frame.getTypeName()
+
+      if functionName
+        tp = as = ''
+        if typeName and functionName.indexOf typeName
+          tp = "#{typeName}."
+        if methodName and functionName.indexOf(".#{methodName}") isnt functionName.length - methodName.length - 1
+          as = " [as #{methodName}]"
+
+        "#{tp}#{functionName}#{as} (#{fileLocation})"
+      else
+        "#{typeName}.#{methodName or '<anonymous>'} (#{fileLocation})"
+    else if isConstructor
+      "new #{functionName or '<anonymous>'} (#{fileLocation})"
+    else if functionName
+      "#{functionName} (#{fileLocation})"
+    else
+      fileLocation
+
+  getSourceMapping = (filename, line, column) ->
+    sourceMap = getSourceMap filename, line, column
+
+    answer = sourceMap.sourceLocation [line - 1, column - 1] if sourceMap?
+    if answer? then [answer[0] + 1, answer[1] + 1] else null
+
+  # Based on [michaelficarra/CoffeeScriptRedux](http://goo.gl/ZTx1p)
+  # NodeJS / V8 have no support for transforming positions in stack traces using
+  # sourceMap, so we must monkey-patch Error to display CoffeeScript source
+  # positions.
+  Error.prepareStackTrace = (err, stack) ->
+    frames = for frame in stack
+      # Don't display stack frames deeper than the stack root `CoffeeScript.run`
+      # for example.
+      break if stackRoot and (frame.getFunction() is stackRoot)
+      "    at #{formatSourcePosition frame, getSourceMapping}"
+
+    "#{err.toString()}\n#{frames.join '\n'}\n"
+
+module.exports = {
+  attach
+}
```

Comment:
Please don’t split this out into a new file. There hasn’t been a new file created out of the main core in many years, probably at least five years, and I don’t want to start now. We’d have to dig into how the build scripts assemble everything, like how certain parts do and don’t go into the browser build, and I don’t want to deal with that now. That’s out of scope for what we’re trying to fix in this PR.

Reply:
It looks like all that is needed for the build is the change in `Cakefile` adding to the browser build list but I also agree that having fewer files can keep things simpler. In that case I think this `attach` will go back to `coffeescript.coffee` and be used from both `register` and the browser docs code. That will also keep it closer to the original implementation.

This is getting very close. It also fixes another bug I found in browser based Try CoffeeScript where the stack trace line numbers are only right for the first time the code runs. Press "Run" twice [here](https://coffeescript.org/#try:fn%20%3D%20-%3E%0A%20%20throw%20new%20Error%20'heyy'%0Afn()%0A) to see the error. The last remaining piece is to hide the inline map from the compiled source in the JS preview editor. The browser needs at least the `//# sourceURL` comment in the executed code to be able to link up to the proper filename for the source map.

Reply:
Thanks. I’m sure it’s probably not that big a deal to refactor the build, but that should at least be its own PR. It would also break all open PRs for the old code that moved. We’ve been toying with the idea of splitting up `nodes.coffee` for years now, moving each class into its own file, but it’s never quite seemed worth the effort in terms of lost history and broken PRs. Basically at this point I consider most of the project architecture decisions “fixed,” in that the cost of changing them is too high when the motivation/availability of contributors is as low as it is. Many of those decisions I would certainly make differently today, like having `lib/` version-controlled, but the project is well-enough designed that I don’t consider anything to be of sufficient hindrance that I feel like I _need_ to make a big change in order to be productive.